### PR TITLE
Enabled gather and scatter for bool tensor

### PR DESCRIPTION
**Pull request by izdeby**
- moving stuff around in order to enable bool. 
- Added implementation of atomicAdd(bool, bool)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -8032,6 +8032,10 @@ def _test_gather(self, cast, test_bounds=True):
         actual = torch.gather(src, 2, idx)
         self.assertEqual(actual, expected, 0)
 
+        # Bool test case
+        t = torch.tensor([[False, True],[True, True]])
+        self.assertEqual(torch.gather(t, 1, torch.tensor([[0, 0], [1, 0]])), torch.tensor([[False, False], [True, True]]))
```
Comment (zou3519) [inline test/test_torch.py]: Needs cuda test case

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 2]
```diff
@@ -443,6 +443,125 @@ void THTensor_(indexFill)(THTensor *tensor, int dim, THLongTensor *index, scalar
   THLongTensor_free(index);
 }
 
+void THTensor_(gather)(THTensor *tensor, THTensor *src, int dim, THLongTensor *index)
```
Comment (VitalyFedyunin) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: Worth mentioning that is pure move, no line changes introduced (at least `git diff --color-moved` thinks so)

File: test/test_torch.py [hunk 3]
```diff
@@ -8032,6 +8032,10 @@ def _test_gather(self, cast, test_bounds=True):
         actual = torch.gather(src, 2, idx)
         self.assertEqual(actual, expected, 0)
 
+        # Bool test case
+        t = torch.tensor([[False, True],[True, True]])
```
Comment (VitalyFedyunin) [inline test/test_torch.py]: lint will complain about missing space

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp (hunk 2 above)
Comment (zou3519) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: Is this just moving the code around in this file?

**Review by VitalyFedyunin — APPROVED** [review]

File: test/test_torch.py (hunk 1 above)
Comment (VitalyFedyunin) [reply test/test_torch.py]: Imo `cast` takes care of it

Comment (VitalyFedyunin) [reply test/test_torch.py]: Ah, wait, it is supposed to take care of it =) but it is missing

Comment (zou3519) [reply test/test_torch.py]: Ah, didn't see that, thanks for pointing that out

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp (hunk 2 above)
Comment (izdeby) [reply aten/src/TH/generic/THTensorEvenMoreMath.cpp]: I added description that its mostly moving stuff around, but couldn't say that its only that as i have added attomicAdd overload.

Comment (izdeby) [reply aten/src/TH/generic/THTensorEvenMoreMath.cpp]: yeah

**Comment by izdeby** [comment]
@VitalyFedyunin, just fyi, i've updated the atomicAdd implementation. It used to call AtomicAddIntegerImpl which was causing a bug: .to(torch.int) would return 2 instead of 1 on CUDA.

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 4]
```diff
@@ -443,6 +443,125 @@ void THTensor_(indexFill)(THTensor *tensor, int dim, THLongTensor *index, scalar
...
+                       TH_TENSOR_DIM_APPLY3_SIZE_SCATTER,
+                       for (i = 0; i < elems_per_row; ++i)
+                       {
+                         idx = *(index_data + i*index_stride);
+                         if (idx < 0 || idx >= tensor_size)
+                         {
+                           THFree(TH_TENSOR_DIM_APPLY_counter);
+                           THError("Invalid index in scatterAdd");
+                         }
+                         tensor_data[idx * tensor_stride] += *(src_data + i*src_stride);
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: is this correct for bool?

Comment (izdeby) [reply aten/src/TH/generic/THTensorEvenMoreMath.cpp]: Why not? Are you concerned that the math might be off for bool? 
I have a separate test case test_scatter_add_bool which is added specifically for this.

**Pull request closed without merging**